### DISCUSSION TITLE
PVS-Studio analysis of spdlog discovered some issues

### DISCUSSION THREAD

**Discussion by MikhailJacques** [gabime/spdlog#3008] [Q&A] [answered]
I have run PVS-studio analysis on my C++ project, including the external sources, one of them being spdlog. The analysis discovered a few issues in several files in spdlog\fmt\bundled directory. While the list of the issues found by PVS-Studio is not exhaustive, I decided to report some. These issues may need to be fixed.

File format-inl.h:

Line 1072: V557 Array underrun is possible. The value of 'offset' index could reach -594. 
Line 1072: V557 Array overrun is possible. The value of 'offset' index could reach 618.
uint64_t pow5 = powers_of_5_64[offset];

Line 223: V610 Unspecified behavior. Check the shift operator '>>'. The left operand '-1' is negative.
static_assert((-1 >> 1) == -1, "right shift is not arithmetic"); 

Line 224: V610 Unspecified behavior. Check the shift operator '>>'. The left operand is negative ('(e * 315653)' = [-827010860..827010860]).
return (e * 1741647) >> 19;

Line 230: V610 Unspecified behavior. Check the shift operator '>>'. The left operand is negative ('(e * 1741647)' = [-2147450751..2147450751]).
return (e * 1741647) >> 19;

Line 234: V610 Unspecified behavior. Check the shift operator '>>'. The left operand is negative ('(e * 631305 - 261663)' = [-1884707088..1853249817]).
return (e * 631305 - 261663) >> 21;

Line 358: V610 Undefined behavior. Check the shift operator '>>'. The right operand is negative ('(32 - beta)' = [-31..31]).
static_cast<uint32_t>(r >> (32 - beta)) == 0};

Line 871: V730 Not all members of a class are initialized inside the constructor. Consider inspecting: ptr_. core.h: buffer(size_t sz) 
noexcept : size_(sz), capacity_(sz) {}

File core.h:

Line 1584: V730 Not all members of a class are initialized inside the constructor. Consider inspecting: type_.
basic_format_arg(const detail::named_arg_info<char_type>* args, size_t size)

Line 2798: V1037 Two or more case-branches perform the same actions. Check lines: 2798, 2804
result.format = float_format::general;

Line 769: V560 A part of conditional expression is always false: 0 >= 1200.
if (detail::is_constant_evaluated() && FMT_GCC_VERSION >= 1200) {

File common-inl.h:

Line 56: V1067 Throwing from an exception constructor may lead to unexpected behavior. Consider to check function call. 
msg_ = fmt::to_string(outbuf);

File format.h:

Line 874: V1077 The 'basic_memory_buffer' constructor contains potentially uninitialized members. Inspect the following: store_.  
FMT_CONSTEXPR20 basic_memory_buffer(basic_memory_buffer&& other) noexcept {

Lines 3674-3679:
V730 Not all members of a class are initialized inside the constructor. Consider inspecting: buffer_.
explicit format_int(int value) : str_(format_signed(value)) {}
explicit format_int(long value) : str_(format_signed(value)) {}
explicit format_int(long long value) : str_(format_signed(value)) {}
explicit format_int(unsigned value) : str_(format_unsigned(value)) {}
explicit format_int(unsigned long value) : str_(format_unsigned(value)) {}
explicit format_int(unsigned long long value)

**Comment by tt4g** [comment]
All sources under the `include/spdlog/fmt/bundled` directory are copied from the [fmt repository](https://github.com/fmtlib/fmt).
They will not be modified in this repository.

**Comment by gabime** [comment, marked as answer]
Please open an issue in the fmt lib repo (https://github.com/fmtlib/fmt)

**Reply by MikhailJacques** [reply]
Just did. Thank you
